test/xbgpu/test_engine.py

```diff
@@ -506,46 +523,49 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
+            for j in range(n_total_accumulations):
+                # The first heap is an incomplete accumulation containing a
+                # single batch, we need to make sure that this is taken into
+                # account by the verification function.
+                if j == 0:
+                    # This is to handle the first accumulation processed. The value
+                    # checked here is simply the first in the range.
+                    # - Even though :meth:`generate_expected_output` returns a
+                    #   zeroed array for an incomplete accumulation, we still need
+                    #   to maintain programmatic sense in the values generated here.
+                    num_batches_in_current_accumulation = HEAPS_PER_FENGINE_PER_CHUNK
                     incomplete_accums_counter += 1
-
-            expected_output = generate_expected_output(
-                base_batch_index,
-                num_batches_in_current_accumulation,
-                heap_accumulation_threshold,
-                n_channels_per_stream,
-                n_ants,
-                n_spectra_per_heap,
-                missing_antenna,
-            )
-            base_batch_index += num_batches_in_current_accumulation
-
-            np.testing.assert_equal(expected_output, device_results[i])
-
-        assert prom_diff.get_sample_diff("output_x_incomplete_accs_total") == incomplete_accums_counter
-        assert prom_diff.get_sample_diff("output_x_heaps_total") == n_total_accumulations
+                else:
+                    num_batches_in_current_accumulation = heap_accumulation_threshold
+                    if missing_antenna is not None:
+                        incomplete_accums_counter += 1
+
+                expected_output = generate_expected_output(
+                    base_batch_index,
+                    num_batches_in_current_accumulation,
+                    heap_accumulation_threshold,
+                    n_channels_per_stream,
+                    n_ants,
+                    n_spectra_per_heap,
+                    missing_antenna,
+                )
+                base_batch_index += num_batches_in_current_accumulation
+
+                np.testing.assert_equal(expected_output, device_results[i][j])
+
+        assert (
+            prom_diff.get_sample_diff("output_x_incomplete_accs_total", {"stream": "bcp1"}) == incomplete_accums_counter
+        )
+        assert prom_diff.get_sample_diff("output_x_heaps_total", {"stream": "bcp1"}) == n_total_accumulations
         # Could manually calculate it here, but it's available inside the send_stream
-        assert prom_diff.get_sample_diff("output_x_bytes_total") == (
-            xbengine.send_stream.heap_payload_size_bytes * n_total_accumulations
+        assert prom_diff.get_sample_diff("output_x_bytes_total", {"stream": "bcp1"}) == (
+            xbengine._pipelines[0].send_stream.heap_payload_size_bytes * n_total_accumulations
```

Comment:
Could use `_request_pipeline` to do the lookup by name, to ensure that you're querying the right pipeline i.e. the one matching `{"stream": "bcp1"}`.

Reply:
I've changed the way these final checks are done to only loop over XPipelines in `xbengine._pipelines` - I use `pipeline.output.name` when looking up stats.